bdsf/multi_proc.py

```diff
@@ -11,28 +11,27 @@
 from __future__ import print_function
 import traceback
 import sys
-
+import os
 import numpy
-_multi = False
+import multiprocessing
 _ncpus = 1
 
-try:
-    # May raise ImportError
-    import multiprocessing
-
-    # Set spawn method to "fork". This is needed for macOS on Python 3.8+ where the
-    # default has been changed to "spawn", causing problems (see the discussion at
-    # https://github.com/ipython/ipython/issues/12396)
-    if sys.platform == 'darwin':
-        if sys.version_info[0] == 3 and sys.version_info[1] >= 8:
-            multiprocessing.set_start_method('fork')
-    _multi = True
-
-    # May raise NotImplementedError
-    _ncpus = min(multiprocessing.cpu_count(), 8)
-
-except:
-    pass
+# Get the number of available cores. We use os.sched_getaffinity() for this if
+# possible, as the number of available cores may be less than the total number
+# of CPU cores in the machine, which is returned by, e.g.,
+# multiprocessing.cpu_count()
+#
+# Note: since macOS (Darwin) does not support os.sched_getaffinity(), we use
+# multiprocessing.cpu_count() instead
+if sys.platform == 'darwin':
+    if sys.version_info[0] == 3 and sys.version_info[1] >= 8:
+        # We need to set spawn method to "fork" for macOS on Python 3.8+ where
+        # the default has been changed to "spawn", causing problems (see the
+        # discussion at https://github.com/ipython/ipython/issues/12396)
+        multiprocessing.set_start_method('fork')
+    _ncpus = multiprocessing.cpu_count()
```

Comment:
This will still give you too many processes on MacOS. I'm not a 100% sure if `os.sched_getaffinity` is available on MacOS. The documentation is silent on that. We could consider to use `len(psutil.Process().cpu_affinity())` as a more portable solution, however, maybe that function is also not supported on MacOS, and we would drag in a dependency on another package (`pustil`) for just a single command.

Reply:
`os.sched_getaffinity` is not available on my mac, python 3.9.13 nor 3.12.5. On MacOS, also `psutil.Process()` does not have a method `cpu_affinity` on MacOS.